lib/charms/opensearch/v0/helper_cluster.py

```diff
@@ -29,53 +29,40 @@ def __init__(self, name: str, roles: List[str], ip: str):
 
 
 class ClusterTopology:
-    """Class for creating the best possible configuration for a Node.
-
-    The current logic is to try to get to the config:
-        - 2 dedicated cluster manager nodes
-        - 1 voting only data node
-    And create them with the following order:
-        - cm0
-        - data, voting only
-        - cm1
-        - data
-        - data
-    """
+    """Class for creating the best possible configuration for a Node."""
 
     @staticmethod
-    def suggest_roles(nodes: List[Node]) -> List[str]:
-        """Get roles for a Node, for now, we only focus on the 3 most important roles.
+    def suggest_roles(nodes: List[Node], planned_units: int) -> List[str]:
+        """Get roles for a Node.
 
-        We will do more interesting things with the nodes list, to find the best role.
+        For now, we don't allow to end-user control roles.
         The logic here is:
-            - the first node should be a CM-eligible node, but will add "data" to it
-            - the second node should be a data and voting-only node, so that:
-                - a 2 nodes cluster can function
-                - when a 2CM-eligible node joins, the CM voting can happen immediately
-            - the 3rd one should be a CM-eligible, but will add "data" to it
-            - the +4 nodes should be data etc.
+            — Half of the nodes should be CM-eligible.
+            — All others should not participate in the voting to speedup voting time.
         """
         nodes_by_roles = ClusterTopology.nodes_count_by_role(nodes)
-        if nodes_by_roles.get("cluster_manager", 0) == 0:
-            return ["cluster_manager", "data"]
 
-        if nodes_by_roles.get("voting_only", 0) == 0:
-            return ["voting_only", "data"]
+        max_managers = planned_units
+        max_voters = planned_units
+        if planned_units % 2 == 0:
+            max_managers -= 1
+            max_voters -= 1
 
-        if nodes_by_roles["cluster_manager"] == 1:
-            return ["cluster_manager", "data"]
+        if max_managers > 3:
+            max_managers = max_managers // 2 + 1
```

Comment:
This seemed confusing to me and not intuitive. It's also possible I'm just being slow.

Are you doing something similar to https://github.com/canonical/kafka-operator/blob/main/src/config.py#L121-L122? I'm biased as I wrote it, but it seemed easier to intuit.

If not applicable, a comment here with business logic would be great.

Reply:
Are you referring to the `//` expression? This is basically the equivalent of `math.ceil(max_managers / 2)`, without the `+1` it would have been a `math.floor`.
And the goal is to have for a cluster of +3 nodes, half the nodes as `cluster_manager` 
Did i manage to clear it up?

Reply:
Oh jesus I blanked and thought it was a Java comment. OK it's fine.

Reply:
Do add a comment though :)

Reply:
Or, better yet, add a new var_name for `max_managers`. `max_managers_needed/halved` is a bad name but you get the idea.